dev-docs/RFCs/v6.1/transform-webgl1-support.md

```diff
@@ -0,0 +1,59 @@
+
+## Main idea
+
+We should be able to setup a render pipeline (strictly using WebGL1 features), such that input parameters are provided as attributes to the VertexShader, inside VertexShader user provided operations are performed, and result is passed to FragmentShader, which writes the result to a pixel as fragment color.
+
+## Implementation details
+
+### Render target
+
+Size of the target should be such that it has one pixel per each expected output. I.e. if user wants to perform an operation on two float array of size 10K , then render target should have at least 10K pixels. Type of render target :
+
+#### Option#1 Framebuffer
+
+We can create a texture and use it as color attachment to a `Framebuffer` object. There might be limitation on some of the browsers using FLOAT texture as render target, for example Safari doesn't support `WEBGL_color_buffer_float` hence a float texture can't be used as color attachment. We might be able to overcome this issue using recently added [pack](https://github.com/uber/luma.gl/blob/master/modules/imageprocessing/src/modules/pack.md) shader module of luma.gl.
+
+#### Option#2 System color buffer
+
+In this option, instead of using Framebuffer object we can use default render color buffer. This has no limitations regarding float type, but can't be used with other graphics applications that render to system provided color buffer. But can be used with applications that do not render anything, like data processing and unit testing etc.
+
+We can start implementation with Option#1.
+
+Once the render target is setup, and data is rendered we can use `gl.readPixels` to read the data.
+
+### Vertex Processing
+
+Idea is to treat each input parameter as an attribute of a vertex and render all vertices as `gl.POINTS`. And perform required operations in Vertex Shader and pass the result into Fragment Shader as a varying. But to be able to place this result onto a single corresponding pixel, we need to also set `gl_Position` value in vertex shader. This position attribute can be internally calculated based on number of elements and set on vertex shader.
+
+### Fragment Processing
+
+Fragment shader will be just a pass through shader, it will move varying value into `gl_FragColor`.
+
+NOTE: Given this effort is to support WebGL1, all shaders must be in GLSL100 syntax.
+
+## Limitations
+
+### Ouput (one vec4)
+
+There can only be one vec4 (four component vector) output. When using WebGL2 TransformFeedback API, there can be more than one varying that can be captured into a Buffer. But given we are using full render cycle to render into color buffer, output is limited to one vec4.
+
+### GPU CPU sync
+
+Under WebGL2, `Transform` allows data to remain on GPU memory and be consumed by the render cycles without syncing with the CPU. But under WebGL1, we need to perform readPixels after every run, resulting in CPU and GPU sync.
```

Comment:
> readPixels after every run, resulting in CPU and GPU sync.

Maybe not, if we support chaining using Texture inputs? see above.

Reply:
This was under the assumption of using default color buffer render target. But yes, when using the texture, we don't have to perform read pixels and just return the texture object.